hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java

```diff
@@ -1867,6 +1870,10 @@ executorService.new ExecutorConfig().setExecutorType(ExecutorType.RS_SNAPSHOT_OP
       choreService.scheduleChore(brokenStoreFileCleaner);
     }
 
+    if (this.rsMobFileCleanerChore != null) {
+      choreService.scheduleChore(rsMobFileCleanerChore);
```

Comment:
So now we will always have this running at RSes, plus the old MobFileCleanerChore running on Master?

Reply:
The new MobFileCleanerChore is not the same as the old one. The main cleanup is done by the RSMobFileCleanerChore running on each RS, cleaning up mob files created by regions hosted on that RS. This should cover most of the usecases.
But I still had to run a restricted version of the old MobFileCleanerChore to clean up after mobs made by now-archived regions, because to make sure those are not referenced anymore every hfile have to be checked.

Reply:
>But I still had to run a restricted version of the old MobFileCleanerChore to clean up after mobs made by now-archived regions, because to make sure those are not referenced anymore every hfile have to be checked.

So, MobFileCleanerChore seems to be checking all existing hfiles. Wouldn't that supersede the RSMobFileCleanerChore work? And what if both run at the same time, could both try to archive a same subset of files?

Reply:
Let me clarify the differences between the 2 cleaners.

RSMobFileCleanerChore:
- runs on RS to have access to currently written files and active storefile list
- can only archive mob files created by regions hosted on the current RS
- only reads hfiles belonging to regions hosted by the current RS when looking for references
This allow it to do the majority of the cleanup necessary as efficiently as possible

MobFileCleanerChore:
- runs on Master
- can only archive mob files created by archived regions (regions no longer existing in the /data folder). Thanks to this these mob files can no longer be "currently written" so we do not need the data only available on the RS
- reads every single hfile in /data with a mob enabled CF. This is necessary, because this is the only way if any of these mobs has active references

So yes, there is an overlap, the same hfiles could be read by both cleansers but the a mob file could be archived with either one of them. The cleaner on the master is wasteful and not especially elegant, but thanks to the lack of centralized mob tracking we do not have a better way of collecting the references.

Reply:
> can only archive mob files created by archived regions (regions no longer existing in the /data folder). Thanks to this these mob files can no longer be "currently written" so we do not need the data only available on the RS

Got it. 

> The cleaner on the master is wasteful and not especially elegant, but thanks to the lack of centralized mob tracking we do not have a better way of collecting the references.

Can we trust META here to list SPLIT/MERGE parents? We could then iterate through these list and check for mob files for these regions. Just brainstorming here, ain't sure if the region dir is left lingering in the FS when SPLIT/MERGE parents are cleared from META (in which case, would make such alternative unfeasible).